src/mpid/ch4/shm/xpmem/xpmem_init.c

```diff
@@ -51,21 +51,46 @@ int MPIDI_XPMEM_mpi_init_hook(int rank, int size, int *n_vcis_provided, int *tag
         MPIDI_XPMEM_global.segmaps[i].apid = -1;        /* get apid at the first communication  */
 
         /* Init AVL tree based segment cache */
-        MPIDI_XPMEM_segtree_init(&MPIDI_XPMEM_global.segmaps[i].segcache);
+        MPIDI_XPMEM_segtree_init(&MPIDI_XPMEM_global.segmaps[i].segcache_user);
+        MPIDI_XPMEM_segtree_init(&MPIDI_XPMEM_global.segmaps[i].segcache_cnt);
     }
 
     /* Check to make sure all processes initialized XPMEM correctly. */
     if (anyfail) {
         for (i = 0; i < num_local; i++) {
-            MPIDI_XPMEM_segtree_delete_all(&MPIDI_XPMEM_global.segmaps[i].segcache);
+            MPIDI_XPMEM_segtree_delete_all(&MPIDI_XPMEM_global.segmaps[i].segcache_user);
+            MPIDI_XPMEM_segtree_delete_all(&MPIDI_XPMEM_global.segmaps[i].segcache_cnt);
         }
         /* Not setting an mpi_errno value here because we can handle the failure of XPMEM
          * gracefully. */
         goto fn_fail;
     }
 
     /* Initialize other global parameters */
+    MPIR_CHKPMEM_MALLOC(MPIDI_XPMEM_global.coop_counter_direct, MPIDI_XPMEM_cnt_t **,
+                        sizeof(MPIDI_XPMEM_cnt_t *) * num_local,
+                        mpi_errno, "xpmem direct counter array", MPL_MEM_SHM);
+    MPIR_CHKPMEM_MALLOC(MPIDI_XPMEM_global.coop_counter_origin_direct, MPIDI_XPMEM_seg_t **,
+                        sizeof(MPIDI_XPMEM_seg_t *) * num_local,
+                        mpi_errno, "xpmem direct counter segs", MPL_MEM_SHM);
+
     MPIDI_XPMEM_global.sys_page_sz = (size_t) sysconf(_SC_PAGESIZE);
+    uint64_t cnt_mem_addr = (uint64_t) (&MPIDI_XPMEM_cnt_mem_direct);
+    uint64_t remote_cnt_mem_addr;
+
+    MPIDU_Init_shm_barrier();
+    MPIDU_Init_shm_put(&cnt_mem_addr, sizeof(uint64_t));
+    MPIDU_Init_shm_barrier();
+    for (i = 0; i < num_local; ++i) {
+        MPIDU_Init_shm_get(i, sizeof(uint64_t), &remote_cnt_mem_addr);
+        mpi_errno = MPIDI_XPMEM_seg_regist(i, sizeof(MPIDI_XPMEM_cnt_t) * MPIDI_XPMEM_CNT_PREALLOC,
+                                           (void *) remote_cnt_mem_addr,
+                                           &MPIDI_XPMEM_global.coop_counter_origin_direct[i],
+                                           (void **) &MPIDI_XPMEM_global.coop_counter_direct[i],
+                                           MPIDI_XPMEM_CNT_BUFFER);
+        MPIR_ERR_CHECK(mpi_errno);
```

Comment:
I'd prefer to directly specify the tree pointer rather than defining extra buffer kind.

Reply:
Can we skip seg_regist for the counters owned by the process itself?

Reply:
I can change it to tree pointer.
For your second comment, I think we have discussed that we can use seg_regist for it. Do you have extra concerns about it right now?

Reply:
It is OK to use `seg_regist`. I meant the process does not need to register its local direct counter.

Reply:
When a process send message to itself, we can use the uniform way to deal with it. I think seg_regist its own local direct counter is necessary and neat design.

Reply:
When a message is sent to the process itself, we should not use coop-copy at all. Single-copy is more efficient. Actually XPMEM should not handle such use case, e.g., no need to attach user buffer. Can you please debug a program to see how the `self-message` is handled?

Reply:
good point, I will let it fall back to single-side copy instead `self-message` case.